File: benchmark/benchmark.rb
#!/usr/bin/env ruby
require 'benchmark'

def fact(n)
     (1..n).inject(1, :*)
end

p fact(20)
p fact(42)

count = 1000

Benchmark.bm do |x|
  x.report("20!/19!") { count.times { fact(20)/fact(19) == 20 } }
  x.report("100!/99!") { count.times { fact(100)/fact(99) == 100 } }
  x.report("1000!/999!") { count.times { fact(1000)/fact(999) == 1000 } }
end

File: benchmark/benchmark.pl
#!/usr/bin/env perl
use v5.18;
use warnings;
use Benchmark qw(:all);
use Math::BigInt    # try => 'GMP';
  warn Math::BigInt->config->{lib};

sub fact {
    my $result = 1;
    $result *= $_ for 1 .. $_[0];
    $result;
}

sub bfact {
    my $result = Math::BigInt->new(1);
    $result->bmul($_) for 1 .. $_[0];
    $result;
}

my $count = shift || 0;

say "20! == ", fact 20;
say "42! == ", bfact 42;

cmpthese timethese $count => {
    'native' => sub { fact(20) / fact(19) == 20   or die },
    'BigInt' => sub { bfact(20) / bfact(19) == 20 or die }
};

cmpthese timethese $count => {
    '20!/19!'  => sub { bfact(20) / bfact(19) == 20   or die },
    '100!/99!' => sub { bfact(100) / bfact(99) == 100 or die }
};
